iceoryx-sys/src/publisher.rs

```diff
@@ -104,38 +105,33 @@ impl Publisher {
         }
     }
 
-    pub fn try_allocate<T>(&self) -> Option<*mut T> {
+    pub fn try_allocate<T>(&self) -> Option<RawSampleMut<MaybeUninit<T>>> {
         let size = std::mem::size_of::<T>() as u32;
         let align = std::mem::align_of::<T>() as u32;
         unsafe {
-            let payload = self.try_allocate_chunk(size, align);
-
-            if !payload.is_null() {
-                Some(payload as *mut T)
-            } else {
-                None
-            }
+            self.try_allocate_chunk(size, align)
+                .map(|payload| payload.cast::<MaybeUninit<T>>())
         }
     }
 
-    pub fn try_allocate_slice<T>(&self, len: u32, align: u32) -> Option<*mut T> {
+    pub fn try_allocate_slice<T>(
+        &self,
+        len: u32,
+        align: u32,
+    ) -> Option<RawSampleMut<[MaybeUninit<T>]>> {
         unsafe {
             if align < std::mem::align_of::<T>() as u32 {
                 return None;
             }
 
             let size = len * std::mem::size_of::<T>() as u32;
-            let payload = self.try_allocate_chunk(size, align);
-
-            if !payload.is_null() {
-                Some(payload as *mut T)
-            } else {
-                None
-            }
+            self.try_allocate_chunk(size, align).map(|payload| {
+                RawSampleMut::slice_from_raw_parts(payload.cast::<MaybeUninit<T>>(), len as usize)
+            })
         }
     }
 
-    unsafe fn try_allocate_chunk(&self, size: u32, align: u32) -> *mut c_void {
+    unsafe fn try_allocate_chunk(&self, size: u32, align: u32) -> Option<RawSampleMut<c_void>> {
```

Comment:
why not use `std::alloc::Layout` here instead of `size: u32, align: u32`?

Reply:
Because it is not user facing API and the checks are repeated in the C++ code base. `std::alloc::Layout` would make the usage slightly more cumbersome for no additional gains. The user facing code interferes these values from generic parameters.

Reply:
... but this is something which might be a good idea to refactor nevertheless